### PR TITLE
Bomb Rush Cyberfunk: Implement new game

### PR DESCRIPTION
**Pull request by TRPG0**
## What is this fixing or adding?
Adds Team Reptile's Bomb Rush Cyberfunk as a new game.

## How was this tested?
Game has been playable and open for anyone to try in its own #future-game-design thread for ~5 months.

### CODE REVIEW

File: worlds/bomb_rush_cyberfunk/__init__.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+        for r, e in region_exits.items():
+            exits = {region_names[ex] for ex in region_exits[r]}
+            self.get_stage(r, player).add_exits(exits)
+
+        for index, loc in enumerate(location_table):
+            stage: Region = self.get_stage(loc["stage"], player)
+            stage.add_locations({loc["name"]: base_id + index})
+            if self.options.skip_polo_photos and "Polo" in loc["name"]:
+                self.options.exclude_locations.value.add(loc["name"])
```
Comment (Exempt-Medic) [inline worlds/bomb_rush_cyberfunk/__init__.py:168]: Is there a reason these are excluded instead of being entirely removed?

Comment (Exempt-Medic) [reply worlds/bomb_rush_cyberfunk/__init__.py:168]: You could also expand the option so the player can decide whether they are excluded or removed entirely.

Comment (TRPG0) [reply worlds/bomb_rush_cyberfunk/__init__.py:168]: I think I'd rather not do that since I feel like having both "Excluded" and "Removed" as options would probably be confusing to people who aren't familiar with Archipelago. With that said though, I think you've changed my mind. I'll make the option remove the locations, and then people who know advanced YAML options can just add them to `exclude_locations` themselves instead if that's what they want.

File: worlds/bomb_rush_cyberfunk/__init__.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    )]
+
+
+class BombRushCyberfunkWorld(World):
+    """Bomb Rush Cyberfunk is 1 second per second of advanced funkstyle. Battle rival crews and dispatch militarized 
+    police to conquer the five boroughs of New Amsterdam. Become All City."""
+
+    game = "Bomb Rush Cyberfunk"
+    web = BombRushCyberfunkWeb()
+    data_version = 1
```
Comment (ScipioWright) [inline worlds/bomb_rush_cyberfunk/__init__.py:29]:
```suggestion
```
Data version is deprecated, and there's been talks about fully removing it. Would recommend deleting this line.

File: worlds/bomb_rush_cyberfunk/__init__.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+        self.multiworld.itempool += pool
+
+
+    def get_stage(self, stage: BRCStage, player: int) -> Region:
+        return self.multiworld.get_region(region_names[stage], player)
+
+
+    def create_regions(self):
+        world = self.multiworld
```
Comment (ScipioWright) [inline worlds/bomb_rush_cyberfunk/__init__.py:153]:
```suggestion
        multiworld = self.multiworld
```
Please do not do world = self.multiworld

File: worlds/bomb_rush_cyberfunk/Items.py [hunk 4]
```diff
@@ -0,0 +1,552 @@
...
+        'type': BRCType.REP},
+
+    # App
+    {'name': "Camera App",
+        'type': BRCType.Camera}
+]
+
+
+group_table: Dict[str, Set[str]] = {
+    "graffitim": ["Graffiti (M - OVERWHELMME)",
```
Comment (ScipioWright) [inline worlds/bomb_rush_cyberfunk/Items.py:410]: The typing says Dict[str, Set[str]], but it appears this is a Dict[str, List[str]]

File: worlds/bomb_rush_cyberfunk/__init__.py (hunk 2 above)
Comment (TRPG0) [reply worlds/bomb_rush_cyberfunk/__init__.py:29]: That's good to know

File: worlds/bomb_rush_cyberfunk/__init__.py (hunk 3 above)
Comment (TRPG0) [reply worlds/bomb_rush_cyberfunk/__init__.py:153]: Noted

File: worlds/bomb_rush_cyberfunk/Items.py (hunk 4 above)
Comment (TRPG0) [reply worlds/bomb_rush_cyberfunk/Items.py:410]: Oops, you're right. I always get the syntax for them confused with each other

File: worlds/bomb_rush_cyberfunk/__init__.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+            self.item_classification[BRCType.BMX] = ItemClassification.progression
+
+
+    def create_items(self):
+        rep_locations: int = 87
+        if self.options.skip_polo_photos:
+            rep_locations -= 18
+
+        self.options.total_rep.round_to_nearest_step()
+        rep_counts = self.options.total_rep.get_rep_item_counts(self.multiworld.random, rep_locations)
```
Comment (ScipioWright) [inline worlds/bomb_rush_cyberfunk/__init__.py:109]:
```suggestion
        rep_counts = self.options.total_rep.get_rep_item_counts(self.random, rep_locations)
```
Recommended to use self.random instead of self.multiworld.random so that things stay deterministic

File: worlds/bomb_rush_cyberfunk/Options.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+                index = random_source.randint(0, location_count-1)
+            items[index] = increment_item(items[index])
+
+        while sum(items) > self.value:
+            index = random_source.randint(0, location_count-1)
+            if not (items[index] == 16 or items[index] == 24 or items[index] == 32):
+                index = random_source.randint(0, location_count-1)
+            items[index] -= 8
+
+        return [items.count(8), items.count(16), items.count(24), items.count(32), items.count(48)]
```
Comment (ScipioWright) [inline worlds/bomb_rush_cyberfunk/Options.py:72]: `get_rep_item_counts` does not consistently give the correct number of items. On default settings, it sometimes gives 86 items instead of 87, leading to one less item being created than there are locations.

File: worlds/bomb_rush_cyberfunk/__init__.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+            self.get_stage(r, player).add_exits(exits)
+
+        for index, loc in enumerate(location_table):
+            if self.options.skip_polo_photos and "Polo" in loc["name"]:
+                continue
+            stage: Region = self.get_stage(loc["stage"], player)
+            stage.add_locations({loc["name"]: base_id + index})
+
+        for e in event_table:
+            stage: Region = self.get_stage(loc["stage"], player)
```
[the remark this answers is not in the document]
Comment (TRPG0) [reply worlds/bomb_rush_cyberfunk/__init__.py:166]: Wow that's pretty bad lol. Guess I never noticed since they must end up in the "Misc" region and the rules are specific enough for it to work like I wanted anyway. Will fix that though

**Review by ScipioWright — APPROVED** [review]
Read through the code, looked fine for the most part. Did lots of test gens, they fixed errors found pretty quickly, and I was unable to find more errors.
Did not test the client -- I'm assuming their beta tests are sufficient for that.

File: worlds/bomb_rush_cyberfunk/docs/en_Bomb Rush Cyberfunk.md [hunk 8]
```diff
@@ -0,0 +1,27 @@
+# Bomb Rush Cyberfunk
+
+## Where is the options page?
+
+The [player options page for this game](../player-options) contains all the options you need to configure and export 
+a config file.
+
+## What does randomization do in this game?
+
+The goal of Bomb Rush Cyberfunk randomizer is to defeat all rival crews in each borough of New Amsterdam, but several
```
Comment (DeamonHunter) [inline worlds/bomb_rush_cyberfunk/docs/en_Bomb Rush Cyberfunk.md:10]: [Optional] Passage could possibly be reordered to read a bit better? Something like
```markdown
## What does randomization do in this game?
The goal of Bomb Rush Cyberfunk randomizer is to defeat all rival crews in each borough of New Amsterdam.

Items can be found by picking up any type of collectible ...

## What changes have been made to the base game?
- The prelude in the police station can be skipped.
...
The mod also adds two new apps to the phone, ...
```

File: worlds/bomb_rush_cyberfunk/docs/setup_en.md [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+1. Download and install [BepInEx 5.4.22 x64](https://github.com/BepInEx/BepInEx/releases/tag/v5.4.22) in your Bomb Rush 
+Cyberfunk root folder. *Do not use any pre-release versions of BepInEx 6.*
+
+2. Start Bomb Rush Cyberfunk once so that BepInEx can create its required configuration files.
+
+3. Download the zip archive from the [releases](https://github.com/TRPG0/BRC-Archipelago/releases) page, and extract its 
+contents into `BepInEx\plugins`.
+
+After installing Archipelago, there are some additional mods that can also be installed for a better experience:
```
Comment (DeamonHunter) [inline worlds/bomb_rush_cyberfunk/docs/setup_en.md:22]: [Optional] Might be best to separate this as another heading so it doesn't feel necessary for setup. Stardew valley lists mods as "Optional Software" as an example.

**Review by DeamonHunter — CHANGES_REQUESTED** [review]
I'm not the most knowledgeable person with Python, so I may be missing some issues. That being said, the main issue is the indentation of several funcs in rules making it hard to parse.

Would also recommend making some tests, in particular:
- A test to ensure that adding/remove rep items works as intended. (Due to collect and remove)
- A test to ensure the correct graffiti count is gotten at various points in the game. (So that refactoring or other changes in that area doesn't break other parts.)

File: worlds/bomb_rush_cyberfunk/Options.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+import typing
+
+if typing.TYPE_CHECKING:
+    from random import Random
+else:
+    Random = typing.Any
+
+
+class Logic(Choice):
+    """Choose the logic used by the randomizer."""
```
[the remark this answers is not in the document]
Comment (TRPG0) [reply worlds/bomb_rush_cyberfunk/Options.py:12]: I don't really feel like this is necessary. In BRC there are pretty much just two glitches (zipping and ceiling clips) that tear this game to shreds, so anyone who's familiar with glitches or speedrunning will already have a good idea of what they're expected to do.

File: worlds/bomb_rush_cyberfunk/docs/en_Bomb Rush Cyberfunk.md (hunk 8 above)
Comment (TRPG0) [reply worlds/bomb_rush_cyberfunk/docs/en_Bomb Rush Cyberfunk.md]: Adjusted it

File: worlds/bomb_rush_cyberfunk/docs/setup_en.md (hunk 9 above)
Comment (TRPG0) [reply worlds/bomb_rush_cyberfunk/docs/setup_en.md:22]: I'm choosing to leave this as is

**Review by DeamonHunter — APPROVED** [review]

File: worlds/bomb_rush_cyberfunk/Rules.py [hunk 11]
```diff
@@ -0,0 +1,1027 @@
...
+        add_rule(multiworld.get_location("Defeat Faux", player),
+            lambda state: rep(state, player, 1000))
+
+    
+    # graffiti spots
+    spots: int = 0
+    while spots < 385:
+        spots += 5
+        set_rule(multiworld.get_location(f"Tagged {spots} Graffiti Spots", player),
+            lambda state: graffiti_spots(state, player, movestyle, limit, glitched, spots))
```
Comment (qwint) [inline worlds/bomb_rush_cyberfunk/Rules.py:1022]: i believe this is making all tagged spots locations out of logic until you can access 380 spots because of how lambdas and variables work in loops
changing to lambda state, spots=spots: should fix that

Comment (TRPG0) [reply worlds/bomb_rush_cyberfunk/Rules.py]: Ah, good to know

File: worlds/bomb_rush_cyberfunk/Rules.py [hunk 12]
```diff
@@ -0,0 +1,1043 @@
...
+    set_rule(multiworld.get_location("Pyramid Island: Maze outfit", player),
+        lambda state: skateboard(state, player, movestyle))
+    if not glitched:
+        add_rule(multiworld.get_location("Pyramid Island: Rise joins the crew", player),
+            lambda state: camera(state, player))
+    set_rule(multiworld.get_location("Pyramid Island: Devil Theory joins the crew", player),
+        lambda state: current_chapter(state, player, 5))
+    if photos:
+        set_rule(multiworld.get_location("Pyramid Island: Polo pile 1", player),
+            lambda state: camera(state, player))
```
Comment (Berserker66) [inline worlds/bomb_rush_cyberfunk/Rules.py:983]:
```suggestion
        set_rule(multiworld.get_location("Pyramid Island: Polo pile 1", player),
            camera)
```
These  all could  be shortened like this, as you're just wrapping the function.

**Review by Berserker66 — APPROVED** [review]
